op-supervisor/supervisor/backend/invalidator/invalidator.go

```diff
@@ -0,0 +1,122 @@
+package invalidator
+
+import (
+	"context"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup/event"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/backend/superevents"
+)
+
+// DatabaseRewinder allows us to tell a db to reset to a specific block
+type DatabaseRewinder interface {
+	// Rewind rewinds all databases for the given chain to the given block
+	Rewind(chain eth.ChainID, headBlock eth.BlockID) error
+	LatestBlockNum(chain eth.ChainID) (num uint64, ok bool)
+}
+
+// ChainResetter allows us to tell a chain processor to reset to a specific block
+type ChainResetter interface {
+	ResetToBlock(ctx context.Context, ref eth.L2BlockRef) error
+}
+
+// Invalidator is responsible for handling invalidation events by coordinating
+// the rewind of databases and resetting of chain processors.
+type Invalidator struct {
+	log      log.Logger
+	emitter  event.Emitter
+	rewinder DatabaseRewinder
+	chains   map[eth.ChainID]ChainResetter
```

Comment:
Rather than using a raw map, which may have racing issues, check out RWMap, which we use for basically all `[chainId]Thing` mappings in the supervisor.